### PR TITLE
feat(sound): join loose amps into a shared rack in the NM/SV designer

### PR DESCRIPTION
**Pull request by jvhtec**
Branched from `main`, clean diff (7 files). Lets techs combine **individually-pictured amps** (the single-amp racks you often get after importing a session) into one physical rack, with a touch-first workflow.

## Summary
- **Affected feature:** Sound → NM/SV amp rack designer.
- Adds a **"Unir amps"** mode: select amps across racks and merge them into one rack (max 3, matching LA-RAK/PLM-RAK capacity).

### Workflow (mobile + desktop)
- **"Unir amps"** toolbar toggle enters join mode.
- Tapping amp cells **selects/deselects** them (checkmark badge); selection is capped at 3 with a toast on overflow.
- A **floating action bar** (bottom-center, thumb-reachable) shows `N/3 seleccionados` with **"Unir en un rack"** (enabled at 2+) and **"Cancelar"**.
- Confirming moves the selected amps into a single rack at the first amp's position, drops any emptied source rack, and nudges a partially-emptied rack aside so nothing overlaps. Amp identity (id / preset / IP) is preserved.
- Join mode suspends block dragging and the editor-tap so selection is reliable on touch; state resets whenever the designer opens or closes.

### Structure
- Core move is a **pure `joinAmpsIntoRack()`** in `layout-utils.ts` (fully unit-tested), driven by a **`useAmpJoin` hook** that owns the mode/selection state. Extracting the hook also kept `AmpRackDesigner.tsx` **under the 800-line file-size budget** (782).

## Risk Assessment
- **Risk level:** low. Client-only, additive UI on an existing tool; no DB/function/schema/money paths. The merge is a pure array transform with no data loss (surplus/leftover amps are retained).

## Impact Checklist
- [x] Migration / realtime / RLS impact — none.
- [x] Performance — client bundle +19.4 kB gzip vs baseline (within budget).

## Test Evidence
- npm run lint — **PASS** (0 errors; repository-existing warnings only). npm run typecheck — **PASS**.
- npm run test:run — **271 files / 1,513 tests passed**.
- Targeted review regressions — **32 tests passed** across layout-utils.test.ts and RackBlockCard.test.tsx, covering right-edge leftover placement, join-mode keyboard selection, and preservation of normal rack keyboard controls.
- npm run governance, npm run test:critical, npm run build, npm run budget:bundle, and npm run cap:sync — **PASS**. Bundle delta: **+21.3 kB gzip**, within budget.
- **E2E (mock-auth, iPhone 13 viewport):** import → 3 single-amp racks → "Unir amps" → tap two amps (checkmarks, 2/3 seleccionados) → "Unir en un rack" → verified the two amps land in one rack, the emptied source is gone, and the third rack is untouched.
- Review follow-ups in ebf30d43 and d13f511c: prevent leftover-rack overlap at the right canvas edge, restore full keyboard operation in join mode, and remove the render-phase drag ref mutation; all five review threads have written responses and are resolved.

## Migration Impact
- **Migration required:** no.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code/session_01RKgvRtgYPA6G3mbB5e3LAT)_

## Summary by CodeRabbit

* **New Features**
  * Added an “Unir amps” mode to combine selected amplifiers into a single rack.
  * Enhanced amp selection within the rack layout with clear visual indicators and keyboard support.
  * Added “Unir en un rack” and “Cancelar” actions, including selection limits and success toasts.
* **Bug Fixes**
  * Improved merge behavior for partial joins, avoiding overlaps and dropping emptied source racks.
  * Disabled desktop drag-to-merge on mobile and while join mode is active.
* **Tests**
  * Added coverage for amp-join, rack-merge, target selection, and join-mode keyboard behavior.

### CODE REVIEW

File: src/components/sound/amplifier-tool/rack-designer/layout-utils.ts [hunk 1]
```diff
@@ -67,6 +67,62 @@ export function blockPixelHeight(block: Pick<RackDesignerBlock, 'amps'>): number
...
+  const result: RackDesignerBlock[] = [];
+  for (const block of blocks) {
+    const remainingAmps = block.amps.filter((amp) => !joinedIds.has(amp.id));
+    if (block.id === anchor.id) {
+      result.push(newBlock);
+      if (remainingAmps.length > 0) {
+        // Leftover amps from the anchor rack shift right so they don't overlap.
+        result.push({
+          ...block,
+          x: Math.min(block.x + BLOCK_WIDTH + 30, CANVAS_WIDTH - BLOCK_WIDTH),
```
[the remark this answers is not in the document]
Comment (jvhtec) [reply src/components/sound/amplifier-tool/rack-designer/layout-utils.ts]: Fixed in f1b51e3d. Leftover anchor racks now move right when space permits and fall back to the left at the canvas boundary. A regression test covers an anchor at CANVAS_WIDTH - BLOCK_WIDTH.

File: src/components/sound/amplifier-tool/rack-designer/RackBlockCard.tsx [hunk 2]
```diff
@@ -32,26 +32,42 @@ interface RackBlockCardProps {
...
 
   const handlePointerDown = (event: React.PointerEvent<HTMLDivElement>) => {
     if (event.pointerType === 'mouse' && event.button !== 0) return;
     event.stopPropagation();
+    // Join mode: a tap on an amp cell toggles its selection; block dragging and
+    // the normal editor tap are suspended so selection is reliable on touch.
+    if (joinMode) {
+      const ampCell = (event.target as HTMLElement).closest('[data-amp-id]') as HTMLElement | null;
+      if (ampCell?.dataset.ampId) onAmpToggle?.(ampCell.dataset.ampId);
+      return;
```
[the remark this answers is not in the document]
Comment (jvhtec) [reply src/components/sound/amplifier-tool/rack-designer/RackBlockCard.tsx:78]: Fixed in f1b51e3d. In join mode, each amp cell is now a focusable pressed button whose Enter/Space action toggles selection; rack editing and movement remain disabled. Tests also verify the normal rack keyboard controls outside join mode.

File: src/components/sound/amplifier-tool/rack-designer/layout-utils.ts [hunk 3]
```diff
@@ -67,6 +67,62 @@ export function blockPixelHeight(block: Pick<RackDesignerBlock, 'amps'>): number
...
+          amps: remainingAmps,
+        });
```
[the remark this answers is not in the document]
Comment (jvhtec) [reply src/components/sound/amplifier-tool/rack-designer/layout-utils.ts]: Fixed in f1b51e3d. Leftover anchor racks now move right when space permits and fall back to the left at the canvas boundary. A regression test covers the right-edge case.

File: src/components/sound/amplifier-tool/rack-designer/useAmpJoin.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    setSelectedAmpIds((prev) => {
+      if (prev.includes(ampId)) return prev.filter((id) => id !== ampId);
+      if (prev.length >= AMPS_PER_RACK) {
+        toast({
+          title: 'Máximo alcanzado',
+          description: `Un rack admite hasta ${AMPS_PER_RACK} amplificadores.`,
+          variant: 'destructive',
+        });
+        return prev;
+      }
```
[the remark this answers is not in the document]
Comment (jvhtec) [reply src/components/sound/amplifier-tool/rack-designer/useAmpJoin.ts:58]: Keeping the three-amp limit intentionally. This PR models physical LA-RAK/PLM-RAK capacity, and the desktop drag-merge path already enforces the same AMPS_PER_RACK limit. Removing only the touch limit would make the interaction paths inconsistent and allow an impossible physical rack.

**Comment by jvhtec** [comment]
@coderabbitai review

File: src/components/sound/amplifier-tool/rack-designer/useRackDragMerge.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    if (!targetId) return;
+    setLayout((prev) =>
+      prev ? { ...prev, blocks: mergeRackIntoRack(prev.blocks, blockId, targetId) } : prev,
+    );
+    onMerged?.(targetId);
+    toast({
+      title: 'Racks unidos',
+      description: 'Los amplificadores se han unido en el rack de destino.',
+    });
+  };
```
[the remark this answers is not in the document]
Comment (jvhtec) [reply src/components/sound/amplifier-tool/rack-designer/useRackDragMerge.ts]: Fixed in d13f511c. The render-phase ref write and unused useRef import are removed; onDragEnd now reads mergeTargetId from its render closure while preserving cleanup, merge, callback, and toast behavior. Targeted tests, ESLint, and typecheck pass.

**Comment by jvhtec** [comment]
@coderabbitai review
